DOCSP-35947: write operations modify documents (#2808)
jenssegers/laravel-mongodb: docs/fundamentals/write-operations.txt

Eloquent's fluent syntax to chain an update operation on methods that
return a Laravel collection object.

This section provides examples of the following update operations:

- :ref:`Update a document <laravel-modify-documents-update-one>`
- :ref:`Update multiple documents <laravel-modify-documents-update-multiple>`

.. _laravel-modify-documents-update-one:

Update a Document Examples
~~~~~~~~~~~~~~~~~~~~~~~~~~

You can update a document in the following ways:

- Modify an instance of the model and save the changes by calling the ``save()``
  method.
- Chain methods to retrieve an instance of a model and perform updates on it
  by calling the ``update()`` method.

The following example shows how to update a document by modifying an instance
of the model and calling its ``save()`` method:

.. literalinclude:: /includes/fundamentals/write-operations/WriteOperationsTest.php
   :language: php
   :dedent:
   :start-after: begin model update one save
   :end-before: end model update one save

When the ``save()`` method succeeds, the model instance on which you called the
method contains the updated values.

If the operation fails, {+odm-short+} assigns the model instance a null value.

> The following example shows how to update a document to retrieve and update

Reviewer: S: Since this example demonstrates how to update documents by chaining methods together:

```suggestion
The following example shows how to chain methods to retrieve and update
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This example shows how to use the ``insert()`` Eloquent method to insert
multiple instances of a ``Concert`` model as MongoDB documents. This bulk
insert method reduces the number of calls your application needs to make
to save the documents.

When the ``insert()`` method succeeds, it returns the value ``1``.

If it fails, it throws an exception.

The example code saves multiple models in a single call by passing them as
an array to the ``insert()`` method:

.. note::

    This example wraps the dates in the `MongoDB\\BSON\\UTCDateTime <{+phplib-api+}/class.mongodb-bson-utcdatetime.php>`__
    class to convert it to a type MongoDB can serialize because Laravel
    skips attribute casting on bulk insert operations.

.. literalinclude:: /includes/fundamentals/write-operations/WriteOperationsTest.php
   :language: php
   :dedent:
   :start-after: begin model insert many
   :end-before: end model insert many

.. _laravel-fundamentals-modify-documents:

Modify Documents
----------------

In this section, you can learn how to modify documents in your MongoDB
collection from your Laravel application. Use update operations to modify

> existing documents or to insert a document if no documents match the search

Reviewer: S: [nit] to avoid repeating "documents":
```suggestion
existing documents or to insert a document if none match the search
```

criteria.

Eloquent lets you specify criteria to retrieve documents, perform updates on

> the returned object collections, and save them.

Reviewer: Q: this description is a little confusing - what is returning the "object collections"? Are you using "documents" and "object collections" interchangeably in this sentence?

Author: I'll replace the sentence with what I think might be less confusing.

I think I meant collection object. I can clarify further by using either the class name or adjective "Eloquent".

This section provides examples of the following update operations:

- :ref:`Update a document <laravel-modify-documents-update-one>`
- :ref:`Update multiple documents <laravel-modify-documents-update-multiple>`

.. _laravel-modify-documents-update-one:

Update a Document Examples
~~~~~~~~~~~~~~~~~~~~~~~~~~

You can update a document in the following ways:

- Modify an instance of the model and save the changes by calling the ``save()``
  method.
- Use the fluent syntax to retrieve a matching document and apply updates to
  it by calling the ``update()`` method or other helper methods.

The following example shows how to update a document by modifying an instance

> of the model and calling its ``save()`` method:

Reviewer: S: not necessary, but I like the detailed description of the code example in lines 86-94. Could be helpful to introduce this code example with more specific info

Author: Is there specific info that you think might be helpful to add? I think the insert example adds type information which sets up the rest of the examples on the page, but could be repetitive and draw away from the focus of the rest of the examples.

Reviewer: I was thinking of specifying which fields / values you're updating (`venue`, `ticketsSold`), but agreed that this could be repetitive and I think it's fine as is!

.. literalinclude:: /includes/fundamentals/write-operations/WriteOperationsTest.php
   :language: php
   :dedent:
   :start-after: begin model update one save
   :end-before: end model update one save

> When the ``save()`` method succeeds, the model instance on which you called the
> method.

Reviewer: I: I think this sentence is incomplete. You could add something like this:
```suggestion
When the ``save()`` method succeeds, the model instance on which you called the
method contains the updated values.
```

Author: Whoops, good catch! Will update.

If the operation fails, {+odm-short+} assigns the model instance a null value.

> The following example shows how to update a document by using fluent syntax to
> retrieve the first match and apply updates to the matching document:

Reviewer: S: I think you can condense this description a little; also, the other mentions of "fluent syntax" use "the", so I'd do the same here
```suggestion
The following example shows how to use the fluent syntax to
retrieve and update the first matching document:
```

Author: Good suggestion, thanks!

.. literalinclude:: /includes/fundamentals/write-operations/WriteOperationsTest.php
   :language: php
   :dedent:
   :start-after: begin model update one fluent
   :end-before: end model update one fluent

.. note::

   The ``orderBy()`` call sorts the results by the ``_id`` field to
   guarantee a consistent sort order. To learn more about sorting in MongoDB,
   see the :manual:`Natural order </reference/glossary/#std-term-natural-order>`
   glossary entry in the {+server-docs-name+}.

When the ``update()`` method succeeds, the operation returns the number of
documents updated.

If the retrieve part of the call does not match any documents in the collection,
{+odm-short+} returns the following error:

.. code-block:: none
   :copyable: false

   Error: Call to a member function update() on null

.. _laravel-modify-documents-update-multiple:

Update Multiple Documents Example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To perform an update on one or more documents, call the ``update()``
method on a collection of model objects. You can use the fluent syntax to

> retrieve a collection of model objects that match your criteria and then

Reviewer: S: since collection usually refers to a MongoDB collection, using that word here (as a synonym of "group", from my understanding) could be misleading. You could change to:
```suggestion
retrieve multiple model objects that match your criteria and then
```

Author: Will clarify the connection between MongoDB documents and a Laravel collection object.

pass your updates to the ``update()`` method.

> The following example shows how to chain calls to retrieve a collection of

Reviewer: S: same comment regarding "collection" as above (unless you are referring to a whole MongoDB collection?)

Author: Thanks for pointing this out. I'll make sure to establish the connection between "documents" and "Laravel collection objects" when necessary, and otherwise use only one of those terms in a paragraph.
